File: routes/student_management.py
"""
Student management (teacher-only) — round-trip CSV siswa.

Endpoints:
    POST /students/export-csv      JSON: {"student_ids": [...]}
    POST /students/import/preview  multipart: file
    POST /students/import          multipart: file
    POST /students/bulk-delete     JSON: {"student_ids": [...]}

Keamanan:
- Autentikasi guru HANYA dari cookie HttpOnly `student_token` — teacher token
  tidak diterima dari query maupun body.
- Validasi Origin untuk semua POST.
- Upload diproses in-memory (tidak pernah ditulis ke /tmp/filesystem).
- Response preview/import/delete TIDAK memuat raw token, token hash, maupun
  exception internal. Token plaintext tidak pernah di-log.
"""

import logging
import os
from datetime import datetime

from flask import Blueprint, Response, jsonify, request

from services import repositories
from services.database import SessionLocal
from services.student_roundtrip import (
    MAX_FILE_BYTES,
    RoundTripImportError,
    mask_student_id,
    parse_roundtrip_csv,
)
from services.token_service import validate_token

student_management_bp = Blueprint("student_management", __name__)

logger = logging.getLogger(__name__)

ALLOWED_EXTENSIONS = {"csv"}


def _teacher_from_cookie():
    """Guru dari cookie HttpOnly `student_token`; None bila tidak terautentikasi."""
    token = request.cookies.get("student_token", "")
    if not token:
        return None
    info = validate_token(token)
    if not info or not info.get("is_teacher"):
        return None
    return info


def _check_origin() -> bool:
    """Origin harus sesuai konfigurasi aplikasi (ORIGIN) bila origin dikirim."""
    allowed = os.environ.get("ORIGIN", "").strip()
    if not allowed or allowed == "*":
        return True
    origin = request.headers.get("Origin", "")
    if not origin:
        return True  # same-origin / non-browser client
    return origin in {allowed, allowed.rstrip("/")}


def _active_lesson_slugs():
    if SessionLocal is None:
        return []
    db = SessionLocal()
    try:
        return [lesson.slug for lesson in repositories.list_active_lessons(db)]
    finally:
        db.close()


def _read_csv_upload():
    """Baca file multipart in-memory. Return (bytes, filename) atau raise ValueError."""
    file = request.files.get("file")
    if file is None or not file.filename:
        raise ValueError("File CSV wajib disertakan (field 'file')")
    filename = file.filename
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""
    if ext not in ALLOWED_EXTENSIONS:
        raise ValueError("Ekstensi file harus .csv")
    # Cek ukuran SEBELUM membaca isi agar upload raksasa tidak di-buffer penuh.
    content_length = getattr(file, "content_length", None)
    if content_length and content_length > MAX_FILE_BYTES:
        raise ValueError(f"File melebihi batas {MAX_FILE_BYTES // (1024 * 1024)} MiB")
    content = file.read()
    if not content:
        raise ValueError("File kosong")
    if len(content) > MAX_FILE_BYTES:
        raise ValueError(f"File melebihi batas {MAX_FILE_BYTES // (1024 * 1024)} MiB")
    return content, filename


@student_management_bp.route("/students/export-csv", methods=["POST"])
def export_students():
    if not _check_origin():
        return jsonify({"success": False, "message": "Origin tidak diizinkan"}), 403
    teacher = _teacher_from_cookie()
    if teacher is None:
        return jsonify({"success": False, "message": "Unauthorized (Teacher only)"}), 401
    if SessionLocal is None:
        return jsonify({"success": False, "message": "PostgreSQL tidak aktif"}), 503

    try:
        data = request.get_json(silent=True, force=True) or {}
        student_ids = data.get("student_ids", []) or []
    except Exception:  # noqa: BLE001
        student_ids = []

    db = SessionLocal()
    try:
        csv_bytes = repositories.export_students_csv(db, student_ids)
    except ValueError as exc:
        db.rollback()
        return jsonify({"success": False, "message": str(exc)}), 400
    except Exception:  # noqa: BLE001 — jangan bocorkan detail internal
        db.rollback()
        logger.exception("Export CSV gagal")
        return jsonify({"success": False, "message": "Gagal mengekspor siswa"}), 500
    finally:
        db.close()

    filename = f"data_siswa_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"
    return Response(
        csv_bytes,
        mimetype="text/csv; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@student_management_bp.route("/students/import/preview", methods=["POST"])
def import_preview():
    if not _check_origin():
        return jsonify({"success": False, "message": "Origin tidak diizinkan"}), 403
    teacher = _teacher_from_cookie()
    if teacher is None:
        return jsonify({"success": False, "message": "Unauthorized (Teacher only)"}), 401
    if SessionLocal is None:
        return jsonify({"success": False, "message": "PostgreSQL tidak aktif"}), 503

    try:
        content, _filename = _read_csv_upload()
    except ValueError as exc:
        return jsonify({"success": False, "message": str(exc)}), 400

    try:
        rows = parse_roundtrip_csv(content, _active_lesson_slugs())
    except RoundTripImportError as exc:
        return jsonify({"success": False, "message": "File tidak valid", "errors": exc.errors}), 400

    db = SessionLocal()
    try:
        summary = repositories.preview_student_import(db, rows)
    finally:
        db.close()

    # Preview memakai nama & masked student_id — TIDAK token/hash.
    preview_rows = [
        {
            "line": row.line,
            "student_id": mask_student_id(row.student_id),
            "nama_siswa": row.display_name,
            "progress_lessons": len(row.progress),
        }
        for row in rows
    ]
    return jsonify({"success": True, "summary": summary, "rows": preview_rows})


@student_management_bp.route("/students/import", methods=["POST"])
def import_students():
    if not _check_origin():
        return jsonify({"success": False, "message": "Origin tidak diizinkan"}), 403
    teacher = _teacher_from_cookie()
    if teacher is None:
        return jsonify({"success": False, "message": "Unauthorized (Teacher only)"}), 401
    if SessionLocal is None:
        return jsonify({"success": False, "message": "PostgreSQL tidak aktif"}), 503

    try:
        content, _filename = _read_csv_upload()
    except ValueError as exc:
        return jsonify({"success": False, "message": str(exc)}), 400

    try:
        rows = parse_roundtrip_csv(content, _active_lesson_slugs())
    except RoundTripImportError as exc:
        return jsonify({"success": False, "message": "File tidak valid", "errors": exc.errors}), 400

    db = SessionLocal()
    try:
        result = repositories.run_student_import(db, rows)
    except RoundTripImportError as exc:
        db.rollback()
        return jsonify(
            {
                "success": False,
                "message": "Import ditolak — ada data siswa/token yang bertentangan dengan database",
                "errors": exc.errors,
            }
        ), 409
    except Exception:  # noqa: BLE001 — jangan bocorkan exception internal
        db.rollback()
        logger.exception("Import CSV gagal")
        return jsonify({"success": False, "message": "Gagal mengimpor siswa"}), 500
    finally:
        db.close()

    logger.info(
        "Import CSV: students_created=%s students_updated=%s "
        "progress_created=%s progress_restored=%s",
        result["students_created"],
        result["students_updated"],
        result["progress_created"],
        result["progress_restored"],
    )
    return jsonify(
        {
            "success": True,
            "students_created": result["students_created"],
            "students_updated": result["students_updated"],
            "progress_created": result["progress_created"],
            "progress_restored": result["progress_restored"],
        }
    )


@student_management_bp.route("/students/bulk-delete", methods=["POST"])
def bulk_delete():
    if not _check_origin():
        return jsonify({"success": False, "message": "Origin tidak diizinkan"}), 403
    teacher = _teacher_from_cookie()
    if teacher is None:
        return jsonify({"success": False, "message": "Unauthorized (Teacher only)"}), 401
    if SessionLocal is None:
        return jsonify({"success": False, "message": "PostgreSQL tidak aktif"}), 503

    try:
        data = request.get_json(silent=True, force=True) or {}
        student_ids = data.get("student_ids", [])
    except Exception:  # noqa: BLE001
        student_ids = []

    db = SessionLocal()
    try:
        deleted_ids = repositories.delete_students(db, student_ids)
        db.commit()
    except ValueError as exc:
        db.rollback()
        return jsonify({"success": False, "message": str(exc)}), 400
    except Exception:  # noqa: BLE001
        db.rollback()
        logger.exception("Bulk delete gagal")
        return jsonify({"success": False, "message": "Gagal menghapus siswa"}), 500
    finally:
        db.close()

    return jsonify(
        {
            "success": True,
            "deleted_count": len(deleted_ids),
            "deleted_ids": deleted_ids,
        }
    )

File: frontend/src/lib/services/studentManagement.ts
/**
 * API client student management (teacher-only round-trip CSV).
 *
 * - Export dikirim sebagai POST JSON → Blob → download client-side.
 * - Import preview/apply memakai multipart (file di-memory, tidak di-upload
 *   plaintext ke mana pun di sisi server).
 * - Response import TIDAK pernah memuat raw token / token hash.
 */

export interface ImportSummary {
	rows: number;
	students_to_create: number;
	students_to_update: number;
	progress_to_create: number;
	progress_to_restore: number;
	progress_to_reset?: number;
	conflicts: string[];
}

export interface ImportPreviewRow {
	line: number;
	student_id: string;
	nama_siswa: string;
	progress_lessons: number;
}

export interface ImportPreviewResponse {
	success: boolean;
	summary: ImportSummary;
	rows?: ImportPreviewRow[];
	message?: string;
	errors?: string[];
}

export interface ImportApplyResponse {
	success: boolean;
	students_created?: number;
	students_updated?: number;
	progress_created?: number;
	progress_restored?: number;
	progress_reset?: number;
	message?: string;
	errors?: string[];
}

export interface BulkDeleteResponse {
	success: boolean;
	deleted_count?: number;
	deleted_ids?: string[];
	message?: string;
}

export interface AddStudentResponse {
	success: boolean;
	student_id?: string;
	nama_siswa?: string;
	message?: string;
	errors?: string[];
}

export interface ExportResult {
	blob: Blob;
	filename: string;
}

export function filenameFromDisposition(disposition: string | null): string {
	if (!disposition) return 'data_siswa.csv';
	const match = /filename="?([^"]+)"?/.exec(disposition);
	return match?.[1] ?? 'data_siswa.csv';
}

/** Download Blob sebagai file (client-side). */
export function triggerBlobDownload(blob: Blob, filename: string) {
	const url = URL.createObjectURL(blob);
	const a = document.createElement('a');
	a.href = url;
	a.download = filename;
	document.body.appendChild(a);
	a.click();
	a.remove();
	URL.revokeObjectURL(url);
}

/** Export siswa (selection) → Blob + filename dari Content-Disposition. */
export async function exportStudentsCsv(
	studentIds: string[],
	customFetch: typeof fetch = fetch
): Promise<ExportResult> {
	const res = await customFetch('/api/students/export-csv', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ student_ids: studentIds })
	});
	if (!res.ok) {
		const data = await res.json().catch(() => null);
		throw new Error(data?.message ?? 'Gagal mengekspor CSV');
	}
	return {
		blob: await res.blob(),
		filename: filenameFromDisposition(res.headers.get('Content-Disposition'))
	};
}

/** Preview import — server memvalidasi tanpa menulis apa pun. */
export async function importPreview(
	file: File,
	customFetch: typeof fetch = fetch
): Promise<ImportPreviewResponse> {
	const form = new FormData();
	form.append('file', file);
	const res = await customFetch('/api/students/import/preview', {
		method: 'POST',
		body: form
	});
	return (await parseResponse(res)) as ImportPreviewResponse;
}

/** Apply import (all-or-nothing: siswa baru dibuat, siswa existing di-restore/update). */
export async function importStudents(
	file: File,
	customFetch: typeof fetch = fetch
): Promise<ImportApplyResponse> {
	const form = new FormData();
	form.append('file', file);
	const res = await customFetch('/api/students/import', {
		method: 'POST',
		body: form
	});
	return (await parseResponse(res)) as ImportApplyResponse;
}

/** Bulk delete siswa terpilih. */
export async function bulkDeleteStudents(
	studentIds: string[],
	customFetch: typeof fetch = fetch
): Promise<BulkDeleteResponse> {
	const res = await customFetch('/api/students/bulk-delete', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ student_ids: studentIds })
	});
	return (await parseResponse(res)) as BulkDeleteResponse;
}

/** Tambah satu siswa langsung (nama + token). */
export async function addStudent(
	namaSiswa: string,
	token: string,
	customFetch: typeof fetch = fetch
): Promise<AddStudentResponse> {
	const res = await customFetch('/api/students/add', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ nama_siswa: namaSiswa, token })
	});
	return (await parseResponse(res)) as AddStudentResponse;
}

/**
 * Parsing respons API: error terstruktur backend (success:false + message/errors)
 * diteruskan apa adanya, sedangkan respons non-JSON (403/404/502 proxy, halaman
 * error HTML, dll) diubah menjadi Error dengan pesan yang informatif — bukan
 * menampilkan pesan generik di UI.
 */
async function parseResponse(res: Response): Promise<unknown> {
	if (res.ok) {
		return res.json();
	}
	const data = await res.json().catch(() => null);
	if (data && typeof data === 'object' && (data as { success?: boolean }).success === false) {
		return data; // error terstruktur dari backend (400/401/403/409/500)
	}
	throw new Error(
		(data as { message?: string } | null)?.message ??
			`Server merespons dengan status ${res.status}`
	);
}
